genet/schedule_elements.py

```diff
@@ -792,46 +1051,732 @@ def plot(self, show=True, save=False, output_dir=''):
-        df = None
-        for service_id, service in self.services.items():
-            _df = service.generate_trips_dataframe(gtfs_day=gtfs_day)
-            if df is None:
-                df = _df
-            else:
-                df = df.append(_df)
-        df = df.reset_index(drop=True)
+        df = self.route_attribute_data(
+            keys=['route_short_name', 'mode', 'trips', 'arrival_offsets', 'departure_offsets', 'ordered_stops', 'id'])
+        df = df.rename(columns={'id': 'route', 'route_short_name': 'route_name'})
+        df['route_name'] = df['route_name'].apply(lambda x: x.replace("\\", "_").replace("/", "_"))
+        df['service'] = df['route'].apply(lambda x: self._graph.graph['route_to_service_map'][x])
+        df['service_name'] = df['service'].apply(
+            lambda x: self._graph.graph['services'][x]['name'].replace("\\", "_").replace("/", "_"))
+        df['ordered_stops'] = df['ordered_stops'].apply(lambda x: list(zip(x[:-1], x[1:])))
+        df['departure_offsets'] = df['departure_offsets'].apply(lambda x: list(map(use_schedule.get_offset, x[:-1])))
+        df['arrival_offsets'] = df['arrival_offsets'].apply(lambda x: list(map(use_schedule.get_offset, x[1:])))
+
+        # expand the frame stop to stop and extract offsets for arrival and departure from these stops
+        stop_cols = np.concatenate(df['ordered_stops'].values)
+        dep_offset_cols = np.concatenate(df['departure_offsets'].values)
+        arr_offset_cols = np.concatenate(df['arrival_offsets'].values)
+        df = DataFrame({
+            col: np.repeat(df[col].values, df['ordered_stops'].str.len())
+            for col in set(df.columns) - {'ordered_stops', 'arrival_offsets', 'departure_offsets'}}
+        ).assign(from_stop=stop_cols[:, 0],
+                 to_stop=stop_cols[:, 1],
+                 departure_time=dep_offset_cols,
+                 arrival_time=arr_offset_cols)
+
+        df['from_stop_name'] = df['from_stop'].apply(
+            lambda x: self._graph.nodes[x]['name'].replace("\\", "_").replace("/", "_"))
+        df['to_stop_name'] = df['to_stop'].apply(
+            lambda x: self._graph.nodes[x]['name'].replace("\\", "_").replace("/", "_"))
+
+        # expand the frame on all the trips each route makes
+        trips = np.concatenate(
+            df['trips'].apply(
+                lambda x: [(k, use_schedule.sanitise_time(time, gtfs_day)) for k, time in x.items()]).values)
+        df = DataFrame({
+            col: np.repeat(df[col].values, df['trips'].str.len())
+            for col in set(df.columns) - {'trips'}}
+        ).assign(trip=trips[:, 0],
+                 trip_dep_time=trips[:, 1]).sort_values(by=['route', 'trip', 'departure_time']).reset_index(drop=True)
+
+        df['departure_time'] = df['trip_dep_time'] + df['departure_time']
+        df['arrival_time'] = df['trip_dep_time'] + df['arrival_time']
+        df = df.drop('trip_dep_time', axis=1)
```

Comment:
something that could have been done in a separate PR.... instead of instantiating all of the services and routes within the Schedule to make a composite of their own trips dataframes, this draws on the new method to generate a dataframe with given keys and expands it to give a DataFrame which gives stop to stop rows with time stamps for all trips.